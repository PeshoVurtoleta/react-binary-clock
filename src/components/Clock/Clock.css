.Clock {
    width: 340px;
    height: 266px;
    margin-top: 4px;
    display: grid;
    grid-template-columns: repeat(6, 1fr);
    grid-template-rows: repeat(5, 1fr);
    grid-auto-flow: column;
    justify-content: center;
    grid-gap: 4px 8px;
}