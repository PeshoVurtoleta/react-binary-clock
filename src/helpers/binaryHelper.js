export const getBinary = digit => (digit >>> 0).toString(2);

export const addZeroPrefixBits = (bin, bits) => {
    while (bin.length < bits) bin = "0" + bin;

    return bin;
};