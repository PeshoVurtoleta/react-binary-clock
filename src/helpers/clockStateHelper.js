export const setTimeAndBinary = ([hours, minutes, seconds], digitPerRows)  => timeToBinary => {
    return {
        time: `${hours}:${minutes}:${seconds}`,
        binary: timeToBinary(hours + minutes + seconds, digitPerRows)
    }
};