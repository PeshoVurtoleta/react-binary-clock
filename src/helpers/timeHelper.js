import {addZeroPrefixBits, getBinary} from "./binaryHelper";

export const getTime = fn => {
    const time = new Date();
    const hours = fn(time.getHours());
    const minutes = fn(time.getMinutes());
    const seconds = fn(time.getSeconds());
    
    return [hours, minutes, seconds];
};

export const addZeroPrefix = time => ('0' + time).slice(-2);

export const timeToBinary = (time, digitRows) => {
    return digitRows.reduce((acc, curr, index) => {
        return acc += addZeroPrefixBits(getBinary(time[index]), curr)
    }, "").split("");
};