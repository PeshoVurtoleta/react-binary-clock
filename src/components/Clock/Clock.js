import React from 'react';

import './Clock.css';

import Square from "../Square/Square";
import Time from "../Time/Time";

import squares from "../../mocks/squares";

import {getTime, addZeroPrefix, timeToBinary} from "../../helpers/timeHelper";
import {setTimeAndBinary} from "../../helpers/clockStateHelper";

export default class Clock extends React.Component {
    delay = 1000;
    digitPerRows = [2, 4, 3, 4, 3, 4];
    timeoutId = null;

    state = {
        time: null,
        binary: []
    };

    tick = () => {
        return setTimeout(() => {
            this.setState(() => ({
                ...setTimeAndBinary(getTime(addZeroPrefix), this.digitPerRows)(timeToBinary)
            }));
        }, this.delay);
    };

    componentDidMount() {
        this.timeoutId = this.tick();
    }

    componentDidUpdate() {
        this.timeoutId = this.tick();
    }

    componentWillUnmount() {
        clearTimeout(this.timeoutId);
    }

    render() {
        const {binary, time} = this.state;

        return (
            <div className="Clock">
                {binary.map((b, index) => (
                        <Square on={b === "1"} key={squares[index].id}
                                column={squares[index].column} row={squares[index].row}
                                number={squares[index].number}
                        />
                    )
                )}
                <Time>{time}</Time>
            </div>
        );
    }
}