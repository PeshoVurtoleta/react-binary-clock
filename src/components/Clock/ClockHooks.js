import React, {useEffect, useState} from "react";
import {addZeroPrefix, getTime, timeToBinary} from "../../helpers/timeHelper";
import Square from "../Square/Square";
import squares from "../../mocks/squares";
import Time from "../Time/Time";

export default function Clock() {
    const delay = 1000;
    const digitPerRows = [2, 4, 3, 4, 3, 4];

    const [time, setTime] = useState();
    const [binary, setBinary] = useState([]);

    useEffect(() => {
        const timer = tick();
        return () => clearTimeout(timer);
    }, [time]);

    const tick = () => {
        return setTimeout(() => {
            const [hours, minutes, seconds] = getTime(addZeroPrefix);

            setTime(`${hours}:${minutes}:${seconds}`);
            setBinary(timeToBinary(hours + minutes + seconds, digitPerRows));
        }, delay);
    };

    return (
        <div className="Clock">
            {binary.map((b, index) => (
                    <Square on={b === "1"} key={squares[index].id}
                            column={squares[index].column} row={squares[index].row}
                            number={squares[index].number}
                    />
                )
            )}
            <Time>{time}</Time>
        </div>
    );
}