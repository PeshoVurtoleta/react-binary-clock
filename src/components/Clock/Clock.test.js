import React from 'react';

import Clock from './Clock';

import {getTime} from "../../helpers/timeHelper";

import {shallow, render, mount} from 'enzyme';

jest.useFakeTimers();

describe(('Should render correctly:'), () => {
    test('Should render correctly:', () => {
        const component = mount(<Clock/>);

        expect(component.exists()).toBe(true);
        expect(component).toMatchSnapshot();
    });

    test('test', () => {
        const component = shallow(<Clock/>);

        expect(component.state()).toEqual({
            time: null,
            binary: []
        });

        jest.advanceTimersByTime(1000);

        expect(component.state().time).not.toBe(null);
        expect(component.state().binary.length).not.toBe(0);
    });

    it('invokes clears timeout on componentDidMount', () => {
        const component = shallow(<Clock/>);

        component.unmount();

        expect(clearTimeout).toHaveBeenCalledTimes(1);
    });
});
