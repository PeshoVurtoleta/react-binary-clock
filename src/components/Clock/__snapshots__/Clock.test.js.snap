// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly: Should render correctly: 1`] = `
<Clock>
  <div
    className="Clock"
  >
    <Time>
      <span
        className="Time"
      />
    </Time>
  </div>
</Clock>
`;
